### PR TITLE
feat(sync): download peer penalization

### PR DESCRIPTION
**Pull request by rkrasiuk**
This PR exposes the peer penalization to the downloaders and implements it for headers & bodies downloaders.

While the network handles the request/response errors (timeouts, decoding errors etc) and penalizes the peers, it has no context about the validity of the response. Upon encountering error during response verification, the downloaders penalize the peer with `ReputationChangeKind::BadMessage` resulting in reputation decrease by `BAD_MESSAGE_REPUTATION_CHANGE` points

### CODE REVIEW

File: crates/stages/src/stages/bodies.rs [hunk 1]
```diff
@@ -238,15 +215,16 @@ impl<D: BodyDownloader, C: Consensus> BodyStage<D, C> {
...
-            let header = header_cursor
-                .seek_exact((block_number, header_hash).into())?
-                .ok_or(DatabaseIntegrityError::Header { number: block_number, hash: header_hash })?
-                .1;
+            let (_, header) = header_cursor.seek_exact((block_number, header_hash).into())?.ok_or(
+                DatabaseIntegrityError::Header { number: block_number, hash: header_hash },
+            )?;
+
             if header.ommers_hash == EMPTY_LIST_HASH && header.transactions_root == EMPTY_ROOT {
+                // TODO: fix this
```
Comment (gakonst) [inline crates/stages/src/stages/bodies.rs:223]: If we indeed move to the new changeset structure let's not forget to add a note that the gaps issue with the returned empty bodies stream is no longer present

**Review by gakonst — APPROVED** [review]

File: crates/stages/src/stages/bodies.rs [hunk 2]
```diff
@@ -106,45 +106,22 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
...
-                    header_hash,
-                ),
-                body: body.transactions,
-                ommers: body.ommers.into_iter().map(|header| header.seal()).collect(),
-            };
-
-            // Pre-validate the block and unwind if it is invalid
-            self.consensus
-                .pre_validate_block(&block)
-                .map_err(|err| StageError::Validation { block: block_number, error: err })?;
```
Comment (gakonst) [inline crates/stages/src/stages/bodies.rs:138]: This was a bug before. If a peer sent us a bad block they could force us to unwind (because the Pipeline will catch the StageError::Validation and unwind). But what we actually want is to detect this error in the downloader, discard that body, and penalize the peer (as done below).

File: crates/net/downloaders/src/bodies/concurrent.rs [hunk 3]
```diff
@@ -72,24 +95,35 @@ impl<C: BodiesClient> ConcurrentDownloader<C> {
...
+        let body = response.remove(0);
+        let block = BlockLocked {
+            header: header.clone(),
+            body: body.transactions,
+            ommers: body.ommers.into_iter().map(|header| header.seal()).collect(),
+        };
+        if self.consensus.pre_validate_block(&block).is_err() {
+            self.client.penalize(peer_id);
+        }
+        Ok(block)
```
Comment (gakonst) [inline crates/net/downloaders/src/bodies/concurrent.rs:111]: There's a bug here, we should return err block, not ok(block)

File: crates/stages/src/stages/bodies.rs [hunk 4]
```diff
@@ -106,45 +106,22 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
...
-        // Cursor used to look up headers for block pre-validation
-        let mut header_cursor = db.cursor::<tables::Headers>()?;
-
         // NOTE(onbjerg): The stream needs to live here otherwise it will just create a new iterator
         // on every iteration of the while loop -_-
         let mut bodies_stream = self.downloader.bodies_stream(bodies_to_download.iter());
         let mut highest_block = previous_block;
-        while let Some((block_number, header_hash, body)) =
+        while let Some(block) =
             bodies_stream.try_next().await.map_err(|err| StageError::Internal(err.into()))?
```
Comment (gakonst) [inline crates/stages/src/stages/bodies.rs:114]: If the bodies stream returns an error (which it will given the fix recommended below), then we should not propagate the error up, because it will cause the pipeline to unwind the stage.

```
whilet let Some(block) = bodies_stream.next() {
      let block = match block {
            Ok(inner) => inner,
           Err(err => error!(err); break; // This should be `break` to avoid gaps in the blocks.
      }
      ...
}
```

Comment (gakonst) [reply crates/stages/src/stages/bodies.rs:114]: This generally feels like the pipeline has a massive footgun because the pipeline is doing `pipeline.run(..).await?` which will "close" the pipeline on any error. This makes me think that we should handle pipeline errors at the pipeline layer? https://github.com/paradigmxyz/reth/blob/ae2fc494c8d8d863f484ab53d92ce6a04bfdf1f4/crates/stages/src/pipeline.rs#L189

cc @onbjerg let's track on separate issue and figure out a good solution here

File: crates/stages/src/stages/bodies.rs (hunk 1 above)
Comment (gakonst) [reply crates/stages/src/stages/bodies.rs:223]: @rkrasiuk pls add the above comment in the code to ensure we don't lose the context around this todo

**Review by gakonst — CHANGES_REQUESTED** [review]

File: crates/net/downloaders/src/bodies/concurrent.rs (hunk 3 above)
Comment (rkrasiuk) [reply crates/net/downloaders/src/bodies/concurrent.rs]: fixed

File: crates/stages/src/stages/bodies.rs (hunk 4 above)
Comment (rkrasiuk) [reply crates/stages/src/stages/bodies.rs]: the local issue is fixed, the pipeline stuff still needs to be handled

File: crates/net/downloaders/src/bodies/concurrent.rs [hunk 5]
```diff
@@ -61,7 +61,6 @@ where
             stream::iter(headers.into_iter().map(|header| {
                 (|| self.fetch_body(header))
                     .retry(ExponentialBackoff::default().with_max_times(self.retries))
-                    .when(|err| err.is_retryable())
```
Comment (gakonst) [inline crates/net/downloaders/src/bodies/concurrent.rs:45]: does this mean that now we retry on all errors?

File: crates/stages/src/stages/bodies.rs [hunk 6]
```diff
@@ -490,7 +532,9 @@ mod tests {
         pub(crate) const GENESIS_HASH: H256 = H256::zero();
 
         /// A helper to create a collection of resulted-wrapped block bodies keyed by their hash.
-        pub(crate) fn body_by_hash(block: &BlockLocked) -> (H256, RequestResult<BlockBody>) {
+        pub(crate) fn body_by_hash(
+            block: &BlockLocked,
+        ) -> (H256, Result<BlockBody, DownloadError>) {
```
Comment (gakonst) [inline crates/stages/src/stages/bodies.rs:537]: Should this be `DownloadResult<T>`?

File: crates/stages/src/stages/bodies.rs [hunk 7]
```diff
@@ -346,29 +366,43 @@ mod tests {
...
 
         // Run the stage
         let rx = runner.execute(input);
 
         // Check that the error bubbles up
         assert_matches!(
             rx.await.unwrap(),
-            Err(StageError::Validation { error: consensus::Error::BaseFeeMissing, .. })
+            Ok(ExecOutput { stage_progress: out_stage_progress, done: false, reached_tip: false })
+                if out_stage_progress == stage_progress
```
Comment (gakonst) [inline crates/stages/src/stages/bodies.rs:405]: So let's keep this as-is for now, and in a follow-up @onbjerg will revert these intio errors and pull the error handling / retrying to the pipeline?

File: crates/stages/src/stages/bodies.rs [hunk 8]
```diff
@@ -346,29 +366,43 @@ mod tests {
...
+                (
+                    b.hash(),
+                    Err(DownloadError::BlockValidation {
+                        hash: b.hash(),
+                        error: consensus::Error::BaseFeeMissing,
+                    }),
+                )
+            })
+            .collect::<HashMap<_, _>>();
+        runner.set_responses(responses);
```
Comment (gakonst) [inline crates/stages/src/stages/bodies.rs:396]: why is this change needed?

File: crates/stages/src/stages/bodies.rs [hunk 9]
```diff
@@ -110,9 +110,24 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
...
+        while let Some(result) = bodies_stream.next().await {
+            let block = match result {
+                Ok(block) => block,
+                Err(err) => {
+                    error!(
+                        "Encountered error downloading block {}. Details: {:?}",
+                        highest_block + 1,
+                        err
+                    );
+                    // Exit the stage early, because
```
Comment (gakonst) [inline crates/stages/src/stages/bodies.rs:122]: unfinished sentence.

File: crates/net/downloaders/src/bodies/concurrent.rs [hunk 10]
```diff
@@ -105,10 +104,14 @@ where
...
-        if self.consensus.pre_validate_block(&block).is_err() {
-            self.client.penalize(peer_id);
+
+        match self.consensus.pre_validate_block(&block) {
+            Ok(_) => Ok(block),
+            Err(error) => {
+                self.client.penalize(peer_id);
+                Err(DownloadError::BlockValidation { hash: header.hash(), error })
+            }
         }
```
Comment (gakonst) [inline crates/net/downloaders/src/bodies/concurrent.rs:114]: great!

File: crates/net/downloaders/src/bodies/concurrent.rs (hunk 5 above)
Comment (rkrasiuk) [reply crates/net/downloaders/src/bodies/concurrent.rs:45]: correct, the timeouts are handled by the network, so only validation and internal errors are left

File: crates/stages/src/stages/bodies.rs (hunk 8 above)
Comment (rkrasiuk) [reply crates/stages/src/stages/bodies.rs:396]: because the consensus checks were moved into the downloader and the `TetsConsensus` isn't actually called in the test downloader. we just set the responses

File: crates/interfaces/src/p2p/downloader.rs [hunk 11]
```diff
@@ -22,3 +33,59 @@ pub trait Downloader: Send + Sync {
     /// The consensus engine
     fn consensus(&self) -> &Self::Consensus;
 }
+
+/// The download result type
+pub type DownloadResult<T> = Result<T, DownloadError>;
+
+/// The downloader error type
+#[derive(Error, Debug, Clone)]
+pub enum DownloadError {
```
Comment (mattsse) [inline crates/interfaces/src/p2p/downloader.rs:42]: let's move to error.rs, together with result type.

File: crates/interfaces/src/p2p/downloader.rs [hunk 12]
```diff
@@ -1,17 +1,28 @@
...
 
 /// A stream for downloading response.
-pub type DownloadStream<T> = Pin<Box<dyn Stream<Item = Result<T, DownloadError>> + Send>>;
+pub type DownloadStream<'a, T> = Pin<Box<dyn Stream<Item = DownloadResult<T>> + Send + 'a>>;
+
+/// Generic download client for peer penalization
+pub trait DownloadClient: Send + Sync + Debug {
+    /// Penalize the peer for responding with a message
+    /// that violates validation rules
+    fn penalize(&self, peer_id: PeerId);
```
Comment (mattsse) [inline crates/interfaces/src/p2p/downloader.rs:17]: this should be more explicit, like: `report_bad_message`

**Review by mattsse — CHANGES_REQUESTED** [review]
smol nits, otherwise lgtm

File: crates/stages/src/stages/bodies.rs (hunk 8 above)
Comment (gakonst) [reply crates/stages/src/stages/bodies.rs:396]: :+1:

File: crates/net/downloaders/src/bodies/concurrent.rs (hunk 5 above)
Comment (gakonst) [reply crates/net/downloaders/src/bodies/concurrent.rs:45]: :+1:

**Review by gakonst — APPROVED** [review]
lgtm pending resolving matt's nits can merge
